### PR TITLE
Merge vsvcs

### PR DESCRIPTION
**Pull request by yuval-k**

### CODE REVIEW

File: projects/gateway/pkg/translator/translator.go [hunk 1]
```diff
@@ -53,16 +58,123 @@ func joinGatewayNames(gateways v1.GatewayList) string {
...
+	// they are invalid.
+	for _, gw := range gateways {
+		bindAddress := fmt.Sprintf("%s:%d", gw.BindAddress, gw.BindPort)
+		bindAddresses[bindAddress] = append(bindAddresses[bindAddress], gw)
+	}
 
+	for addr, gateways := range bindAddresses {
+		if len(gateways) > 1 {
+			for _, gw := range gateways {
+				resourceErrs.AddError(gw, fmt.Errorf("bind-addres %s is not unique in a proxy. gateways: %s", addr, strings.Join(gatewaysRefsToString(gateways), ",")))
```
Comment (rickducott) [inline projects/gateway/pkg/translator/translator.go]: bind-address

File: projects/gateway/pkg/translator/translator.go [hunk 2]
```diff
@@ -74,23 +186,34 @@ func desiredListener(gateway *v1.Gateway, virtualServices v1.VirtualServiceList,
 		}
 	}
 
-	var (
-		virtualHosts []*gloov1.VirtualHost
-		sslConfigs   []*gloov1.SslConfig
-	)
-
+	var ret v1.VirtualServiceList
```
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: nit: useful name here in place of `ret`, e.g. `mergedVirtualServices`

File: projects/gateway/pkg/translator/translator.go [hunk 3]
```diff
@@ -74,23 +186,34 @@ func desiredListener(gateway *v1.Gateway, virtualServices v1.VirtualServiceList,
...
 	for _, ref := range virtualServicesForGateway {
 		// virtual service must live in the same namespace as gateway
 		virtualService, err := virtualServices.Find(ref.Strings())
 		if err != nil {
 			resourceErrs.AddError(gateway, err)
```
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: can we wrap this `err` with a message giving some context? e.g. `virtual service %v must live in same namespace as gateway`

File: projects/gateway/pkg/translator/translator.go [hunk 4]
```diff
@@ -53,16 +58,123 @@ func joinGatewayNames(gateways v1.GatewayList) string {
...
+			Metadata:  ref,
+		}
+		ret = append(ret, mergedVs)
+	}
+
+	return ret
+}
+
+func getVirtualServiceForGateway(gateway *v1.Gateway, virtualServices v1.VirtualServiceList, resourceErrs reporter.ResourceErrors) v1.VirtualServiceList {
 	virtualServicesForGateway := gateway.VirtualServices
```
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: `virtualServiceRefs`

File: projects/gateway/pkg/translator/translator.go (hunk 2 above)
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: `virtualServicesForGateway`

File: projects/gateway/pkg/translator/translator.go [hunk 5]
```diff
@@ -53,16 +58,123 @@ func joinGatewayNames(gateways v1.GatewayList) string {
...
+	}
+
+	// report errors
+	for domain, domainSetKeys := range domainSet {
+		if len(domainSetKeys) > 1 {
+			resourceErrs.AddError(gateway, fmt.Errorf("domain %s is present in more than one vservice set in this gateway", domain))
+		}
+	}
+	// return merged list
+	var ret v1.VirtualServiceList
```
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: nit: rename `ret` to `mergedVirtualServices`

File: projects/gateway/pkg/translator/translator.go [hunk 6]
```diff
@@ -53,16 +58,123 @@ func joinGatewayNames(gateways v1.GatewayList) string {
...
+				}
+			} else if havePlugins {
+				resourceErrs.AddError(gateway, fmt.Errorf("more than one vhost plugin is present in virtual service of these domains: %s", k))
+			}
+		}
+		glooutils.SortRoutesByPath(routes)
+
+		ref := core.Metadata{
+			// name shouldnt matter as it this object is ephemeral.
+			Name:      k,
```
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: can we make name something useful for debugging? e.g. `"merged-vs-for-"+k`

File: projects/gateway/pkg/translator/translator.go [hunk 7]
```diff
@@ -53,16 +58,123 @@ func joinGatewayNames(gateways v1.GatewayList) string {
...
+
+			havePlugins := vs.VirtualHost != nil &&
+				vs.VirtualHost.VirtualHostPlugins != nil
+
+			if vhostPlugins == nil {
+				if havePlugins {
+					vhostPlugins = vs.VirtualHost.VirtualHostPlugins
+				}
+			} else if havePlugins {
+				resourceErrs.AddError(gateway, fmt.Errorf("more than one vhost plugin is present in virtual service of these domains: %s", k))
```
Comment (ilackarms) [inline projects/gateway/pkg/translator/translator.go]: i'm not opposed to erroring here for now, however it seems fragile / possibly something we want to support

for example one set of routes may use auth, while another set uses rate limiting. i'm not sure it should be invalid (in the long run)

Comment (yuval-k) [reply projects/gateway/pkg/translator/translator.go]: i think in the whole run the api should be restructured to reflect more of a hierarchy.

File: projects/gateway/pkg/translator/translator.go (hunk 6 above)
Comment (yuval-k) [reply projects/gateway/pkg/translator/translator.go]: this name may be used later in envoy for stats, so i prefer to keep short and clean.

Comment (yuval-k) [reply projects/gateway/pkg/translator/translator.go]: hmm i do have an issue with empty names.. let me see what i can do here

**Pull request merged**